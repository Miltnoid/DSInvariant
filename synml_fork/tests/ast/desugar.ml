type nat =
  | O
  | S of nat

type bool =
  | True
  | False

type src =
  | NumS of nat
  | PlusS of src * src
  | MinusS of src * src
  | TrueS
  | FalseS

type opcode =
  | PlusOp
  | MinusOp

type dst =
  | NumD of nat
  | BinopD of opcode * dst * dst
  | BoolD of bool


let desugar : src -> dst |>
{
  NumS(0) => NumD(0)
| NumS(1) => NumD(1)
| PlusS(NumS(0), NumS(1)) => BinopD(PlusOp, NumD(0), NumD(1))
| PlusS(NumS(1), NumS(0)) => BinopD(PlusOp, NumD(1), NumD(0))    
| MinusS(NumS(0), NumS(1)) => BinopD(MinusOp, NumD(0), NumD(1))
| MinusS(NumS(1), NumS(0)) => BinopD(MinusOp, NumD(1), NumD(0))
| TrueS => BoolD(True)
| FalseS => BoolD(False)
} = ?
                       
