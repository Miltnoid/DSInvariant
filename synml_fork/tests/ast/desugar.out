(* Generated Wed Dec 31 11:17:18 2014 *)
let desugar : src -> dst =
  let rec f1 (s1:src) : dst =
    match s1 with
      | NumS n1 -> NumD n1
      | PlusS (s2, s3) -> BinopD (PlusOp, f1 s2, f1 s3)
      | MinusS (s2, s3) -> BinopD (MinusOp, f1 s2, f1 s3)
      | TrueS -> BoolD True
      | FalseS -> BoolD False
  in
    f1
;;
